client-misk-dynamodb/src/main/kotlin/app/cash/backfila/client/misk/dynamodb/internal/DynamoDbBackfillOperator.kt

```diff
@@ -0,0 +1,93 @@
+
+class DynamoDbBackfillOperator<I : Any, P : Any>(
+  val dynamoDb: DynamoDBMapper,
+  override val backfill: DynamoDbBackfill<I, P>,
+  val parametersOperator: BackfilaParametersOperator<P>
+) : BackfillOperator {
+  override fun name(): String = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config =
+        parametersOperator.constructBackfillConfig(request.parameters, request.dry_run)
+    // TODO(mikepaw): dynamically select the segment count by probing DynamoDB.
+    val segmentCount = backfill.fixedSegmentCount(config) ?: 2048
+    val partitionCount = backfill.partitionCount(config)
+    val segmentsPerPartition = segmentCount / partitionCount
+
+    val partitions = mutableListOf<PrepareBackfillResponse.Partition>()
+    for (i in 0 until partitionCount) {
+      val segmentStartInclusive = i * segmentsPerPartition
+      val segmentEndExclusive = (i + 1) * segmentsPerPartition
+      partitions += PrepareBackfillResponse.Partition.Builder()
+          .partition_name("$i of $partitionCount")
+          .backfill_range(encodeKeyRange(segmentStartInclusive, segmentEndExclusive, segmentCount))
+          .build()
+    }
+
+    return PrepareBackfillResponse.Builder()
+        .partitions(partitions)
+        .build()
+  }
+
+  override fun getNextBatchRange(request: GetNextBatchRangeRequest): GetNextBatchRangeResponse {
+    var (start, end, count) = decodeKeyRange(request.backfill_range)
+
+    // If this isn't the first batch range, start where we last left off.
+    if (request.previous_end_key != null) {
+      start = decodeSegment(request.previous_end_key).first
+    }
+
+    val batches = mutableListOf<GetNextBatchRangeResponse.Batch>()
+    for (i in start until minOf(start + 1, end)) {
+      batches += GetNextBatchRangeResponse.Batch.Builder()
+          .batch_range(encodeKeyRange(i, i + 1, count))
+          .matching_record_count(100L)
+          .scanned_record_count(100L)
+          .build()
+    }
+
+    return GetNextBatchRangeResponse.Builder()
+        .batches(batches)
+        .build()
+  }
+
+  override fun runBatch(request: RunBatchRequest): RunBatchResponse {
+    val (start, end, count) = decodeKeyRange(request.batch_range)
+    require(end == start + 1)
+
+    val config =
+        parametersOperator.constructBackfillConfig(request.parameters, request.dry_run)
+
+    var lastEvaluatedKey: Map<String, AttributeValue>? = null
+    do {
+      val scanRequest = DynamoDBScanExpression().apply {
+        segment = start
+        totalSegments = count
+        limit = 4
+        if (lastEvaluatedKey != null) {
+          exclusiveStartKey = lastEvaluatedKey
+        }
+      }
+      val result = dynamoDb.scanPage(backfill.itemType, scanRequest)
+      backfill.runBatch(result.results, config)
+      lastEvaluatedKey = result.lastEvaluatedKey
+    } while (lastEvaluatedKey != null)
```

Comment:
why do you need to loop here? A batch can span multiple pages? how long would each batch take?

Reply:
So this is where this is a basic version of the backfill. Yes a batch can span multiple pages. With this implementation it could yeah take a really long time.

We have three improvements:
1. During the prepare we are going to try and guess the number of segments to break up the backfill into a bit better. 
2. I'm going to add a way for run batch to essentially partially run, return with "call me again" information that backfila will just call again. Which would complete the next n-items. 
3. This is more risky but we will have an option of establishing batches that are parts of a segment. The contract with DyanamoDB indicates that it should work but since it isn't explicit it's a bit risky. This will use marker row information. If the marker row gets deleted then the backfill will fail. Resuming the backfill will work though.

Reply:
If it wan't clear these will be coming in later PRs.